ARIA attributes

a-d-w-s opened the pull request:
ARIA attributes FormHelper.php

a-d-w-s committed:
ARIA attributes

ndm2 commented on src/View/Helper/FormHelper.php:
```diff
@@ -766,7 +766,10 @@ public function error(string $field, $text = null, array $options = []): string
-        return $this->formatTemplate('error', ['content' => $error]);
+        return $this->formatTemplate('error', [
+            'content' => $error,
+            'id' => $field . '-aria',
```
This value needs to be passed through `FormHelper::_domId()`, same for the `aria-describedby` option value.

And maybe this should have a more generic suffix, so that it doesn't indicate such a strictly tied purpose.

shadowx-jb replied:
I don't understand how I should do it? For ARIA it is necessary that the element has an ID, and it must have some prefix - because the text field itself has an ID. So if they had the same ID, it would not be valid according to W3C.

ndm2 replied:
Since IDs aren't something specifically related to ARIA, I'd suggest to not use `-aria` as a suffix, but something more generic that fits what the element represents, like `-error`.

ndm2 commented on src/View/Helper/FormHelper.php:
```diff
@@ -1046,6 +1049,10 @@ public function control(string $fieldName, array $options = []): string
             'templates' => [],
             'templateVars' => [],
             'labelOptions' => true,
+            'aria-label' => empty($options['placeholder']) ? $options['label'] : $options['placeholder'],
+            'aria-required' => $options['required'] == true ? 'true' : null,
+            'aria-invalid' => $this->isFieldError($fieldName) ? 'true' : null,
+            'aria-describedby' => $this->isFieldError($fieldName) ? $fieldName . '-aria' : null,
```
Generally, do these ARIA attributes maybe need to _not_ apply to hidden inputs?

markstory replied:
I don't think they're required or useful for hidden inputs. Aria agents will skip hidden inputs by default.

shadowx-jb replied:
You're right, I didn't think of hidden fields. It probably shouldn't matter, but I'd rather do it so that ARIA isn't in the hidden fields.

shadowx-jb replied:
I modified the condition, now ARIA will be inserted into everything outside the hidden field and if it has a label or placeholder attribute.

a-d-w-s committed:
Update ARIA attributes

ADmad commented:
These changes need to target `4.next` instead of `master`.

shadowx-jb commented:
> 
> 
> These changes need to target `4.next` instead of `master`.

shadowx-jb commented:
I rebase to `4.next` and push

a-d-w-s committed:
Rename suffix - aria to error

ndm2 commented:
> I rebase to `4.next` and push

You also need to change the target branch of the PR so that it points to `4.next` (click on "edit" in the top right, then you can change the branch).

ndm2 commented on src/View/Helper/FormHelper.php:
```diff
@@ -1044,6 +1047,19 @@ public function control(string $fieldName, array $options = []): string
+                'aria-label' => !empty($options['label']) ? $options['label'] : $options['placeholder'],
+                'aria-required' => $options['required'] == true ? 'true' : null,
+                'aria-invalid' => $isFieldError ? 'true' : null,
+                'aria-describedby' => $isFieldError ? $options['id'] . '-error' : null,
```
Using `$options['id']` won't work, `id` has already been passed through `_domId()`, which transforms the value to include only valid characters, and also might prepend a custom prefix, so it could be different from what is being set in the `error()` method.

You'll need to explicitly pass the field name + suffix through `_domId()`, and do the same in the `error()` method so that both values match.

shadowx-jb replied:
OK, thank You

a-d-w-s committed:
Add $this->_domId to field

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1044,6 +1047,19 @@ public function control(string $fieldName, array $options = []): string
         $options = $this->_parseOptions($fieldName, $options);
         $options += ['id' => $this->_domId($fieldName)];
 
+        if (
+            !empty($options['label']) && $options['type'] != 'hidden' ||
+            !empty($options['placeholder']) && $options['type'] != 'hidden'
+        ) {
```
These empty() calls could eat `'0'` which *could* be a label or placeholder. Doing a comparison on the input type should be enough as an empty aria-label won't hurt.

```suggestion
        if ($options['type'] != 'hidden') {
```

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1044,6 +1047,19 @@ public function control(string $fieldName, array $options = []): string
+            $options += [
+                'aria-label' => !empty($options['label']) ? $options['label'] : $options['placeholder'],
+                'aria-required' => $options['required'] == true ? 'true' : null,
+                'aria-invalid' => $isFieldError ? 'true' : null,
```
```suggestion
                'aria-invalid' => $isFieldError ? true : null,
```

You should be able to use a boolean here. If not we can fix up the attribute generation code.

shadowx-jb replied:
It should be true or false:
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-invalid_attribute

markstory commented on src/View/Helper/FormHelper.php:
```diff
@@ -1044,6 +1047,19 @@ public function control(string $fieldName, array $options = []): string
+        ) {
+            $isFieldError = $this->isFieldError($fieldName);
+            $options += [
+                'aria-label' => !empty($options['label']) ? $options['label'] : $options['placeholder'],
```
The `label` option can also be an array. We'll need to work around that.
```suggestion
                'aria-label' => isset($options['label']) && is_string($options['label']) ? $options['label'] : $options['placeholder'],
```

markstory reviewed (commented):
Looking good. I'll add some tests in a day or so.

a-d-w-s committed:
Update src/View/Helper/FormHelper.php

Co-authored-by: Mark Story <mark@mark-story.com>

a-d-w-s committed:
Update src/View/Helper/FormHelper.php

Co-authored-by: Mark Story <mark@mark-story.com>

shadowx-jb commented:
Question: if we have one error, it writes the text to the div element. If we have more errors, then we write to the div element UL LI. For a well-styled voice - for example, adding an icon in front of the text in case of one error - I would suggest wrapping one error in the p element, which is inserted into the div element. Example: 

```
            } else {
                $error = array_pop($error);
            }
```

New 

```
            } else {
                $error = $this->Html->para(null, array_pop($error));
            }
```

markstory commented:
> I would suggest wrapping one error in the p element, which is inserted into the div element. Example:

While I agree with you, this is more complicated for us to change at the framework level as switching tags out has a higher chance of breaking userland applications as they could be relying on specific tag `names` in their CSS or JavaScript code.

markstory committed:
Fix errors and add first draft of tests.

Some problems I'm noticing.

* aria-label is generally empty. Unless the user provides a label string
  it will be empty.
* aria-describedby doesnt' always exist and we can end up with invalid
  references.

markstory commented on tests/TestCase/View/Helper/FormHelperTest.php:
```diff
@@ -2433,8 +2437,13 @@ public function testErrorMessageDisplay(): void
+                'class' => 'form-error',
+                'aria-invalid' => 'true',
+                // TODO this seems wrong as the id doesn't exist.
+                'aria-describedby' => 'article-title-error',
```
There are a few tests I marked with TODO as the results look off to me.

ndm2 replied:
The `control()` call should probably disable the attribute, ie pass `'aria-describedby' => null`.

ndm2 commented on tests/TestCase/View/Helper/FormHelperTest.php:
```diff
@@ -2911,8 +2964,13 @@ public function testControlCustomization(): void
+                'id' => 'field',
+                'class' => 'form-error',
+                'aria-invalid' => 'true',
+                // TODO this seems bad as the id doesn't exist.
+                'aria-describedby' => 'field-error',
```
Maybe hardcode the `id` attribute in the custom `error` template? The `aria-describedby` option could receive a custom value too, after all that's something that should be tested anyways.

markstory committed:
Fix problems with aria-describedby and remove aria-label

The aria-describedby attribute can only be set properly when we know
that the error text will display and that the error message will have an
id attribute set.

I've also removed the aria-label attribute. Our forms use labels and the
`for` attribute which is enough for assitive technologies to provide
useful feedback.

markstory commented:
I've fixed the problems with `aria-describedby` and removed the `aria-label` attribute. The `aria-label` attribute was only set when an explicit label string was set which might not be very often. Furthermore, the aria-label attribute is redundant when we also have id attributes on inputs and labels with correct `for` attributes set.

shadowx-jb commented:
`aria-label` should be, if I have a `placeholder`, for example ... blind people then don't know what kind of field it is. I use a field, without a label and I have the field title in the placeholder.

markstory commented:
>  I use a field, without a label and I have the field title in the placeholder.

Hasn't the developer gone out of their way to make accessibility bad by not providing a label though? I don't know if the framework should be trying to compensate for that.

shadowx-jb commented:
Accessibility for the blind is that the label is in the `aria-label`. For others, he is a `placeholder `.

markstory commented:
So would you recommend adding an `aria-label` attribute whenever the `placeholder` is set or only if the `placeholder` is defined and the label has been disabled?

shadowx-jb commented:
Yes, if label disabled and placeholder are defined, I would set the aria-label field.

markstory committed:
Use placeholder to create aria-label if the label is disabled.

markstory committed:
Fixed phpcs.
